docs/advanced/optimizing-selectors.md

```diff
@@ -0,0 +1,110 @@
+# Optimizing Selectors
+
+[Selectors](../concepts/select.md) are responsible for providing state data to your application. As your application code grows, naturally the number of selectors you create also increases. Ensuring your selectors are optimized can be instrumental in building a faster performing application.
+
+## Memoization
+
+Selectors are memoized functions. Memoized functions are calculated when their arguments change and the results are cached. Regardless of how many components or services consume a selector, a selector will calculate only once when state changes and the cached result will be returned to all consumers. Taking advantage of this feature can result in performance increases.
+
+For example, there exists this state model:
+
+```ts
+interface SomeStateModel {
+  data: Data[];
+  name: string;
+}
+```
+
+And in this example there is an input component where a user can type a name. On key down, an action is dispatched updating the `name` property of state. On the same page, another component renders `data`. In order to render state data we create a selector in our state class:
+
+```ts
+@Selector()
+static getViewData(state: SomeStateModel) {
+   return state.data.map(d => expensiveFunction(d));
+}
+```
+
+Selectors defined in state classes implicitly have `state` injected as their first argument. The above selector will be recalculated every time the user types into the input component. Since `state` could update rapidly when a user types, the expensive selector will needlessly recalculate even though it does not care about the `name` property of `state` changing. This selector does not take advantage of memoization.
+
+One way to solve this problem is to turn off the `injectContainerState` selector option. This enables you to explicitly specify arguments using the `@Selector` decorator. For example:
```

Comment:
Good call.  That additional information will definitely make things clearer.

Reply:
After some more thought, I think its more correct to convey the following (I'm still working out the verbiage):

This option does not affect the behavior of selectors declared outside of state classes.
This option only affects selectors declared within state classes by preventing state being implicitly injected as the first argument.  This requires you to explicitly specify all arguments using the `@Selector([...])` decorator _that are declared in state classes_.

Motivation:
> Maybe one aside note here is that this setting does not apply to parameterless selectors eg. @Selector().

Since it's possible to have parameterless selectors in state classes., e.g.
```ts
@State({ ... })
@Injectable()
class State {
``` 
Turning the option off, those selectors could still benefit from removing state